### PR TITLE
Ps 1291 fix extension icon updates

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Adds icon updating to manifest V3

Also extracts badges in general to a separate concern from main background and context menus.

## Code changes

- **background.ts**: Add listeners for tabs api and generic message listening.
- **main.background**: separate out badge refreshing and menu refreshing. Call the update-badge methods to set all icons
- **runtime.background/tabs.background**: method name updates.
- **browserApi**: manifest version helper
- **update-badge**: lift from main.background all of the sets badge background, icon, and overlay text each time it's called for both badge and sidebar (if it exists)
- **services.module**: Expand private mode to needsBackgroundInit
- **browserPlatformUtils**: use property for window -- needed for ServiceWorkers

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: apps/browser/src/background/service_factories/two-factor-service.factory.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ export async function twoFactorServiceFactory(
   cache: { twoFactorService?: AbstractTwoFactorService } & CachedServices,
   opts: TwoFactorServiceInitOptions
 ): Promise<AbstractTwoFactorService> {
-  const service = await factory(
```
Comment (justindbaur) [inline apps/browser/src/background/service_factories/two-factor-service.factory.ts:21]: Does this service not actually need to be inited or are we removing that responsibility from the factories?

File: apps/browser/src/background/service_factories/storage-service.factory.ts [hunk 2]
```diff
@@ -38,7 +39,7 @@ export function memoryStorageServiceFactory(
   opts: MemoryStorageServiceInitOptions
 ): Promise<AbstractStorageService> {
   return factory(cache, "memoryStorageService", opts, async () => {
-    if (chrome.runtime.getManifest().manifest_version == 3) {
+    if (BrowserApi.manifestVersion == 3) {
```
Comment (justindbaur) [inline apps/browser/src/background/service_factories/storage-service.factory.ts:42]: nit: Strict equals

File: apps/browser/src/browser/browserApi.ts [hunk 3]
```diff
@@ -201,4 +212,16 @@ export class BrowserApi {
...
+
+  static getBrowserAction() {
+    return BrowserApi.manifestVersion === 3 ? chrome.action : chrome.browserAction;
+  }
+
+  static getSidebarAction(win: Window | typeof global) {
+    return navigator.userAgent.indexOf(" Safari/") !== -1 // is Safari
+      ? null
+      : typeof (win as any).opr !== "undefined" && opr.sidebarAction
+      ? (win as any).opr.sidebarAction
```
Comment (justindbaur) [inline apps/browser/src/browser/browserApi.ts:224]: I don't _think_ you need to `win as any` for `opr` stuff now.

File: apps/browser/src/browser/browserApi.ts [hunk 4]
```diff
@@ -201,4 +212,16 @@ export class BrowserApi {
...
+      : (win as any).chrome.sidebarAction;
```
Comment (justindbaur) [inline apps/browser/src/browser/browserApi.ts:225]: Here either I believe.

File: apps/browser/src/decorators/session-sync-observable/session-syncer.ts [hunk 5]
```diff
@@ -29,7 +29,7 @@ export class SessionSyncer {
   }
 
   init() {
-    if (chrome.runtime.getManifest().manifest_version != 3) {
+    if (BrowserApi.manifestVersion != 3) {
```
Comment (justindbaur) [inline apps/browser/src/decorators/session-sync-observable/session-syncer.ts:32]: nit: Strict equals

File: apps/browser/src/listeners/update-badge.ts [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+type BadgeUpdateType = {
+  setBadgeText: (textDetails: { text: string; tabId: number }) => Promise<void>;
+  setIcon: (iconDetails: { path: any; windowId?: number }, callback?: () => void) => void;
+  setBadgeBackgroundColor: (colorDetails: { color: string; tabId?: number }) => Promise<void>;
+};
+
+type SideBarType = {
+  setTitle: (titleDetails: { title: string; tabId: number }) => Promise<void>;
+};
```
Comment (justindbaur) [inline apps/browser/src/listeners/update-badge.ts:28]: Hopefully these can utilize what's in [here](https://github.com/bitwarden/clients/blob/master/apps/browser/src/globals.d.ts).

**Review by justindbaur — CHANGES_REQUESTED** [review]

File: apps/browser/src/background/service_factories/two-factor-service.factory.ts (hunk 1 above)
Comment (MGibson1) [reply apps/browser/src/background/service_factories/two-factor-service.factory.ts:21]: not quite sure why these changes were included, we should be init-ing. Thanks

**Review by justindbaur — DISMISSED** [review]

**Comment by MGibson1** [comment]
Ready to go from QA. Needs merging for safari tests

**Review by justindbaur — APPROVED** [review]
LGTM

**Pull request merged**